### PR TITLE
Login

### PR DESCRIPTION
**Pull request by gavrilenco**

### CODE REVIEW

File: back/myapp/constants/common.js [hunk 1]
```diff
@@ -18,6 +15,12 @@ const MAX = "2017-12-31"
 const MALE = "male"
 const FEMALE = "female"
 
+const PATH_TO_USER_DATA = "./public/usersData/";
+const SERVER_DOMAINL = "http://localhost:3001";
+const PATH_TO_USER_DATA_C = "/usersData/";
```
Comment (gavrilenco) [inline back/myapp/constants/common.js]: PATH_TO_USER_DATA_C what is C?

File: back/myapp/constants/common.js [hunk 2]
```diff
@@ -18,6 +15,12 @@ const MAX = "2017-12-31"
 const MALE = "male"
 const FEMALE = "female"
 
+const PATH_TO_USER_DATA = "./public/usersData/";
+const SERVER_DOMAINL = "http://localhost:3001";
```
Comment (gavrilenco) [inline back/myapp/constants/common.js]: SERVER_DOMAINL what is L?

File: back/myapp/constants/common.js [hunk 3]
```diff
@@ -18,6 +15,12 @@ const MAX = "2017-12-31"
...
+const PATH_TO_USER_DATA_C = "/usersData/";
+const FULL_PATH_TO_USER_DATA = SERVER_DOMAINL + PATH_TO_USER_DATA_C;
+const PATH_TO_DEFAULT_AVATAR = SERVER_DOMAINL + PATH_TO_USER_DATA_C + "default_avatar.png"
```
Comment (gavrilenco) [inline back/myapp/constants/common.js]: why no to use FULL_PATH_TO_USER_DATA here?

File: back/myapp/constants/errors.js [hunk 4]
```diff
@@ -17,7 +17,11 @@ const RECT_FIELD_ERROR = "rect fields value must be from 0 to 1"
 const FAILED_SEND_MESSAGE_TO_EMAIL = "failed send message to email"
 const SAVE_FILE_ERROR = "save file error"
 const REGISTRATION_USER_MODEL_INVALIDE = " registration user model invalidate"
-
+const UNAUTHORIZED = "unauthorized"
+const INCORECT_DATA_FOR_LOGIN = "incorect data for login"
```
Comment (gavrilenco) [inline back/myapp/constants/errors.js]: incorect -> incorrect

File: back/myapp/routes/index.js [hunk 5]
```diff
@@ -22,26 +30,15 @@ var storage = multer.diskStorage({
...
-        .use(function(err,req,res,next){
-            console.log("ERROR", err)
-            clearRegistrationRequest(req);
-            res.send(res.message);
-            return;
-          });
+router.post(API_METHODS_PATHS.LOGIN, [checkDbConnection, login])
+        .use(simpleErrorHandler);
+        
+router.post(API_METHODS_PATHS.GET_AUTHORIZE_USER_DATA, verifyToken, [checkDbConnection, getAuthUserData])
```
Comment (gavrilenco) [inline back/myapp/routes/index.js]: why POST for getting user data?

File: back/myapp/scripts/account/account.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+var { LOGIN } = require("../../constants/errors");
+var { updateImagePaths } = require("../utils")
+var { User } = require("../../models/User");
+var { dispatchError } = require("../errorHandlers/common")
+var jwt = require("jsonwebtoken");
+
+
+function getAuthUserData(req, res, next){
+    jwt.verify(req.token, loginConfig.secretKey, (err, authData) => {
```
Comment (gavrilenco) [inline back/myapp/scripts/account/account.js]: why we use tokenVerify in other places but here direct call?

File: back/myapp/scripts/account/login.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+var { SIGSTKFLT } = require('constants');
+var { User } = require('../../models/User');
+var jwt = require('jsonwebtoken');
+var { LOGIN } = require('../../constants/errors');
+var { dispatchError } = require("../errorHandlers/common")
+
+function login (req,res,next) {
+    var { email, password } = req.body
+
+    if(typeof email == "undefined")
```
Comment (gavrilenco) [inline back/myapp/scripts/account/login.js]: if (email) will check email != null && email != undefined

File: back/myapp/scripts/account/login.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    {
+      dispatchError(res,next,LOGIN.INCORECT_DATA_FOR_LOGIN,403);
+    }
+
+    User.findOne({email}, (err, user) => {
+      if(user == null){
+        dispatchError(res,next,LOGIN.INCORECT_EMAIL_FOR_LOGIN,403);
```
Comment (gavrilenco) [inline back/myapp/scripts/account/login.js]: if (user == null){
dispatch
return
}

if (user.checkPassword)

File: back/myapp/scripts/errorHandlers/registration.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+var { clearRegistrationRequest } = require("../account/registration");
+
+
+
+function registrationError(err,req,res,next){
+    clearRegistrationRequest(req);
+    res.send(res.message);
```
Comment (gavrilenco) [inline back/myapp/scripts/errorHandlers/registration.js]: you have the same logic in simpleErrorHandler, what is for then?

File: back/myapp/scripts/midllewares/checkDbConnection.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+var { USER_ERRORS } = require("../../constants/errors");
+var mongoose = require("mongoose");
+
+var { LOGIN } = require("../../constants/errors");
+
+function checkDbConnection(req, res, next){
+    if(mongoose.connection.readyState == 0){
+        res.statusCode = 500;
+        res.message = USER_ERRORS.DB_NOT_CONNECTED;
+        next(res.message)
```
Comment (gavrilenco) [inline back/myapp/scripts/midllewares/checkDbConnection.js]: dispatchError?

File: back/myapp/scripts/midllewares/token.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+var { LOGIN } = require("../../constants/errors");
+
+function verifyToken(req, res, next){
+    const bearerHeader = req.headers['authorization'];
+    if(typeof bearerHeader !== 'undefined'){
```
Comment (gavrilenco) [inline back/myapp/scripts/midllewares/token.js]: if (bearerHeader)

File: back/myapp/scripts/users.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+var { loginConfig } = require("../constants/config");
+var { LOGIN } = require("../constants/errors");
+var { User } = require("../models/User");
+var { getPathToImages, updateImagePaths } = require("./utils")
+var jwt = require("jsonwebtoken");
+var { dispatchError } = require("./errorHandlers/common")
+
+
+function getUserData(req, res, next){
+    jwt.verify(req.token, loginConfig.secretKey, (err, authData) => {
```
Comment (gavrilenco) [inline back/myapp/scripts/users.js]: why not tokenVerify?

File: back/myapp/scripts/utils.js [hunk 13]
```diff
@@ -28,6 +28,23 @@ function deleteFiles(files){
   }
 }
 
+function getPathToImages({minAvatar, avatar, _id}){
+  var pathToImage = `${paths.FULL_PATH_TO_USER_DATA}${_id.toString(16)}/` 
+  return {
+    minAvatar: (minAvatar !== null && minAvatar !== undefined) ? pathToImage + minAvatar : paths.PATH_TO_DEFAULT_AVATAR,
```
Comment (gavrilenco) [inline back/myapp/scripts/utils.js]: minAvatar ? code : code

File: back/myapp/scripts/utils.js [hunk 14]
```diff
@@ -28,6 +28,23 @@ function deleteFiles(files){
...
+    avatar: (avatar !== null && avatar !== undefined) ? pathToImage + avatar : paths.PATH_TO_DEFAULT_AVATAR
+  }
+}
+
+function updateImagePaths(user){
```
Comment (gavrilenco) [inline back/myapp/scripts/utils.js]: not sure this is the best name for this function

File: back/myapp/scripts/account/login.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+      } else if(user.checkPassword(password)) {
+        jwt.sign({user}, loginConfig.secretKey,{ expiresIn: "30 days" }, (err, token) => {
```
Comment (gavrilenco) [inline back/myapp/scripts/account/login.js]: FYI: better to have refresh token

File: back/myapp/scripts/users.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+      if(err) {
+          dispatchError(res,next,LOGIN.UNAUTHORIZED,403);
+      }
+      else {
+        var { id } = req.body;
+        User.findById(id, "firstname surname email avatar minAvatar birthdate parrentname", (err, result) => {
```
Comment (gavrilenco) [inline back/myapp/scripts/users.js]: the same request with the same field list I've seen before, probably you should keep the list of fields somewhere

File: back/myapp/scripts/users.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+var { User } = require("../models/User");
+var { getPathToImages, updateImagePaths } = require("./utils")
+var jwt = require("jsonwebtoken");
+var { dispatchError } = require("./errorHandlers/common")
+
+
+function getUserData(req, res, next){
+    jwt.verify(req.token, loginConfig.secretKey, (err, authData) => {
+      if(err) {
+          dispatchError(res,next,LOGIN.UNAUTHORIZED,403);
```
Comment (gavrilenco) [inline back/myapp/scripts/users.js]: if (err){
dispatch
return
}

File: front/social/src/App.js [hunk 18]
```diff
@@ -1,26 +1,33 @@
 import React, { Component } from 'react';
 import ErrorWindow from "./components/common/errorWindow";
 import LoaderWindow from "./components/common/loadingWindow"
-import { Route } from "react-router-dom"
+import { Route, Redirect } from "react-router-dom"
 
 import Welcome from "./components/pages/helloWorld.js"
```
Comment (gavrilenco) [inline front/social/src/App.js]: Welcome still used on this page?

File: front/social/src/actions/Account.js [hunk 19]
```diff
@@ -64,4 +66,72 @@ export const registration = () => (dispatch, getState) => {
...
+    password,
+    isValid
+   } = getState().login;
+
+   if(isValid == false)
+    return;
+
+    dispatch({
+      type: ACTION_FOR_APP.SHOW_LOADING_WINDOW,
+      payload: "Идет авторизация"
```
Comment (gavrilenco) [inline front/social/src/actions/Account.js]: should be in constants

File: front/social/src/actions/Account.js [hunk 20]
```diff
@@ -64,4 +66,72 @@ export const registration = () => (dispatch, getState) => {
...
+
+export const getAuthUserData = () => (dispatch, getState) => {
+  const {
+    token,
+    isAuthorize
+   } = getState().app;
+
+   if(isAuthorize == false)
+    return;
+    console.log("Получение данных авторизованного пользователя")
```
Comment (gavrilenco) [inline front/social/src/actions/Account.js]: console.logs should be removed

File: front/social/src/App.js [hunk 21]
```diff
@@ -1,26 +1,33 @@
...
 
 class App extends Component {
   
   render() {
-    var { loadingWindow } = this.props.app;
+    var { loadingWindow, isAuthorize, authorizedUser } = this.props.app;
     var { errorWindow} = this.props.catcher;
     var { onCloseErrorWindow } = this.props;
+
+    const menu = () => { return (authorizedUser !== null) ? <MainMenu/> : "" };
```
Comment (gavrilenco) [inline front/social/src/App.js]: is this a HOC withAuthorize? this logic should be moved to MainMenu component

File: front/social/src/components/common/header.js [hunk 22]
```diff
@@ -38,16 +43,49 @@ class Header extends React.Component {
...
             </div>
             <div>
-                <Link to="/">
-                    <Button color="inherit">
-                        {content.ToHome}
-                    </Button>
-                </Link>
-                <Link to="/registration">
+              {
+                (()=>{
```
Comment (gavrilenco) [inline front/social/src/components/common/header.js]: looks like it's a separate component (() => {})()

File: front/social/src/components/common/header.js [hunk 23]
```diff
@@ -38,16 +43,49 @@ class Header extends React.Component {
...
+                    </Link>
+                  </div>)
+                  }else {
+                    return (<div className="header-menu-wrapper">
+                        <Typography className="header-name">{authorizedUser.firstname}</Typography>
+                        <div className="header-avatar" onClick={this.handleMenu}>
+                          <CircleAvatar imageUrl={authorizedUser.minAvatar}/>
+                        </div>
+                        <Menu
+                          style={{marginTop:"2.8em"}}
```
Comment (gavrilenco) [inline front/social/src/components/common/header.js]: inline styles?

File: front/social/src/components/common/mainMenu.js [hunk 24]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class MainMenu extends React.Component {
+
+    onLogoutClick = () => {
+      this.props.onLogout();
+    }
+  
+    render() {
+      const { isAuthorize, authorizedUser } = this.props.app;
```
Comment (gavrilenco) [inline front/social/src/components/common/mainMenu.js]: where is isAuthorize used?

File: front/social/src/components/pages/profile/avatar.js [hunk 25]
```diff
@@ -0,0 +1,20 @@
+import {AppBar, Button, Toolbar, Menu ,Typography} from '@material-ui/core';
+import React, { Component } from 'react';
+import Paper from '@material-ui/core/Paper/Paper';
+
+
+class Avatar extends React.Component {
+
+    render() {
+      var { user } = this.props;   
+      var forRender = (user !== null) ? <img src={user.minAvatar} /> : <div className="avatar-emitter" />
```
Comment (gavrilenco) [inline front/social/src/components/pages/profile/avatar.js]: user ? code : code

File: front/social/src/components/pages/profile/mainInfo.js [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+import Grid from '@material-ui/core/Grid/Grid';
+import { PROFILE_CONSTS } from '../../../constans/profile';
+import { PROFILE_CONTENT } from '../../../content/profile';
+
+
+class MainInfo extends React.Component {
+  
+    render() {
+        var { user } = this.props;   
+        var forRender = (user !== null) ? (<MainInfoWithUser user={user}/>) : ( <MainInfoEmitter/>)
```
Comment (gavrilenco) [inline front/social/src/components/pages/profile/mainInfo.js:12]: var why not let or const? is this es6?

File: front/social/src/reducers/login.js [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+                        isError: true,
+                        message: MESSAGE.USER_NOT_FOUND_ABOUT_EMAIL
+                    }
+                }
+            }
+
+        if(data == errors.INCORECT_PASSWORD_FOR_LOGIN)
+            return {
+                ...state,
+                validateState: {
```
Comment (gavrilenco) [inline front/social/src/reducers/login.js]: looks like we need separate reducer for loginValidate

File: front/social/src/reducers/catcher.js [hunk 28]
```diff
@@ -10,6 +10,27 @@ const initialState = {
...
+        errorWindow: {
+            isVisible: true,
+            message: COMMON_MESSAGE.UNEXPECTED_ERROR_MESSAGE
+        }
+    }
+}
+
+
+function getTechnicalState(state){
+    return {
```
Comment (gavrilenco) [inline front/social/src/reducers/catcher.js]: copy paste

File: front/social/src/components/pages/profile/profile.js [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+      this.update(this.props.id);
+    }
+
+    componentWillReceiveProps(nextProps){
+      if(nextProps.id !== this.props.id)
+        this.update(nextProps.id);
+    }
+
+    update(id){
+      var {props} = this;
```
Comment (gavrilenco) [inline front/social/src/components/pages/profile/profile.js]: destructurization?

File: front/social/src/components/pages/profile/mainInfo.js [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+    }
+}
+
+
+class MainInfoEmitter extends React.Component {
+
+    render() {
+        return (
+            <Paper className="main-info-wrapper">
+                sdfsdf
```
Comment (gavrilenco) [inline front/social/src/components/pages/profile/mainInfo.js]: ?

File: front/social/src/components/pages/profile/mainInfo.js [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+        )
+    }
+  }
+
+class MainInfoWithUser extends React.Component {
+
+   render() {
+        var { user } = this.props;
+        var items = [
+            { key: PROFILE_CONTENT.BIRTHDATE, value: new Date(user.birthdate).toLocaleDateString("ru-RU", PROFILE_CONSTS.DATE_FORMAT)}
```
Comment (gavrilenco) [inline front/social/src/components/pages/profile/mainInfo.js]: desctructurization?

File: front/social/src/components/common/mainMenu.js [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+        <div className="main-menu">
+         <Paper className="main-menu-paper">
+            <MenuList>
+                <Link to={`/${authorizedUser._id}`}>
+                    <MenuItem >
+                        <ListItemText primary="Моя страница" />
+                    </MenuItem>
+                </Link>
+                <MenuItem >
+                    <ListItemText primary="Друзья" />
```
Comment (gavrilenco) [inline front/social/src/components/common/mainMenu.js]: constants? how will you implement i18n?

File: back/myapp/scripts/users.js (hunk 16 above)
Comment (yanAnciferov) [reply back/myapp/scripts/users.js]: i think that in future this requests will have different field list

**Pull request merged**